### DISCUSSION TITLE
API parity VS modern API

### DISCUSSION THREAD

**Discussion by MrJul** [AvaloniaUI/Avalonia#9978] [General]
Hello,

I wanted to open a new PR with commit https://github.com/MrJul/Avalonia/commit/dec8c1c586be564857dea712c36851918267fe80 that further improves text layout performance.
Before doing so I had a look at the git history of the changed files, and I now feel it's better to discuss the changes first.

In the above commit, I completely removed `CharacterBufferReference` and `CharacterBufferRange`. My first thoughts when stumbling upon those structs were "oh, some old code ported from WPF, better use the modern and performant `ReadOnlyMemory/ReadOnlySpan` structs instead". The same thing happened with `GlyphRun._characters` being a list: "let's make that a `ReadOnlyMemory`!".

After making most of these changes and being happy with the perf wins and (in my opinion) cleaner API, I noticed that `CharacterBufferReference/Range` were in fact introduced very recently: about a month ago in PR #9464. In that same PR, `_characters` was changed from a `ReadOnlySlice` to a `IReadOnlyList`.

I assume that these changes were made to ease porting applications from WPF to Avalonia.

As a long time WPF developer, I'm a fan of Avalonia: both frameworks share the same concepts, but most of the little frustrating things and limitations from WPF are gone. The API is cleaner and feels modern. Plus there's a real focus on performance. Kudos to the project owners/maintainers and the many contributors!

Please continue to do what you did well: evaluate what's right and wrong in WPF, WinUI, MAUI, etc. and bring the good parts over then make them even better. I'm all for feature parity, but in my opinion bringing the existing APIs as is isn't the solution.

It's not just about performance, which is a hard topic requiring constant effort and that can be easily regressed without noticing it. For what it's worth, I believe `CharacterBufferReference/Range` are genuinely bad: the range isn't a range, but only a length with the starting index being stored in the reference. Slicing the range modifies both the range and the reference, which is quite odd and harder to reason about than the `ReadOnlySlice` that was present before.

Note that my changes don't bring back `ReadOnlySlice`: only the standard `ReadOnlyMemory/Span` are used, which should be even easier to understand as they're widespread types.

Anyway, I don't want to "almost revert" a previous PR without discussing it beforehand. 

What's the correct way of action here: 
 - Is it OK for me to open the PR as is?
 - Should I rework it to keep the `CharacterBufferRange/Buffer` as a public API but use `Memory/Span` internally where possible?
 - Should I simply forget about it?

Overall, what's the philosophy regarding porting API from other frameworks?

I want to improve Avalonia by making further performance pull requests, and knowing how much I'm allowed to change would really help! I'm also happy to discuss all this in a more appropriate medium if needed :)

----

Here are the numbers for the latest `HugeTextLayout` benchmark on my machine:

Before #9464 (`ReadOnlySlice`):
|                Method |       Mean |    Error |   StdDev |    Gen 0 |    Gen 1 | Allocated |
|---------------------- |-----------:|---------:|---------:|---------:|---------:|----------:|
|       BuildTextLayout |   459.4 us |  0.90 us |  0.80 us |  14.1602 |   5.8594 |    232 KB |
| BuildEmojisTextLayout |   927.4 us |  3.84 us |  3.40 us |  37.1094 |  18.5547 |    621 KB |
|   BuildManySmallTexts | 3,738.0 us | 27.43 us | 25.66 us | 300.7813 | 148.4375 |  4,942 KB |
|  VirtualizeTextBlocks | 5,087.2 us | 21.25 us | 18.83 us | 359.3750 |  54.6875 |  5,967 KB |

After #9464 (`CharacterBufferRange`):
|                Method |       Mean |    Error |   StdDev |    Gen 0 |    Gen 1 | Allocated |
|---------------------- |-----------:|---------:|---------:|---------:|---------:|----------:|
|       BuildTextLayout |   795.6 us |  4.50 us |  4.21 us |  63.4766 |  21.4844 |      1 MB |
| BuildEmojisTextLayout | 1,731.6 us |  9.25 us |  8.20 us | 138.6719 |  68.3594 |      2 MB |
|   BuildManySmallTexts | 4,880.1 us | 30.92 us | 28.92 us | 414.0625 | 203.1250 |      7 MB |
|  VirtualizeTextBlocks | 6,277.7 us | 52.45 us | 49.06 us | 484.3750 |  78.1250 |      8 MB |

Current master:
|                Method |       Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Allocated |
|---------------------- |-----------:|---------:|---------:|---------:|--------:|----------:|
|       BuildTextLayout |   634.2 us |  5.67 us |  5.30 us |   0.9766 |       - |     25 KB |
| BuildEmojisTextLayout | 1,364.1 us | 12.15 us | 11.37 us |        - |       - |     30 KB |
|   BuildManySmallTexts | 4,043.9 us | 21.54 us | 20.15 us | 156.2500 | 78.1250 |  2,563 KB |
|  VirtualizeTextBlocks | 5,088.5 us | 20.43 us | 19.11 us | 148.4375 | 23.4375 |  2,445 KB |

After my changes (`ReadOnlyMemory`):
|                Method |       Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Allocated |
|---------------------- |-----------:|---------:|---------:|---------:|--------:|----------:|
|       BuildTextLayout |   355.3 us |  0.95 us |  0.89 us |   0.9766 |       - |     22 KB |
| BuildEmojisTextLayout |   735.4 us |  2.74 us |  2.56 us |   0.9766 |       - |     27 KB |
|   BuildManySmallTexts | 3,267.1 us | 10.25 us |  9.59 us | 144.5313 | 70.3125 |  2,391 KB |
|  VirtualizeTextBlocks | 4,291.4 us | 14.17 us | 12.56 us | 132.8125 | 15.6250 |  2,273 KB |

**Comment by maxkatz6** [comment]
IIRC we needed CharacterBufferRange/Buffer specifically to make API closer to WPF and simplify porting of some new features. Like, rich edit box or others.
CC @Gillibald @Mikolaytis

**Comment by Mikolaytis** [comment]
I want best performance possible, also I want Avalonia to have FlowDocument and RichTextBox asap.
I think that performance is over compatibility in this case.

**Comment by Gillibald** [comment]
Performance difference most likely comes from typed collection vs IReadOnlyList. I am open for moving to ReadOnlySlice/Memory. When I ported CharacterBufferReference I wanted to remove the usage of ReadOnlySlice.Start/End to be able to combine buffers from different sources.

So please share your code and we can make this change happen

**Reply by MrJul** [reply]
> Performance difference most likely comes from typed collection vs IReadOnlyList

Yes, that's where most of the perf improvement comes.

> I am open for moving to ReadOnlySlice/Memory.

Nice, considering `CharacterBufferRange.Skip` and `CharacterBufferReference.ctor` still appeared at hotspots in profiling. We can't get much faster than `Span`! :) (and `Memory` is a necessity, not everything can be a ref struct.)

**Comment by Gillibald** [comment]
https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.trygetstring?view=net-7.0 is a game changer hier wasn't aware of this API

This allows avoiding tracking buffer offsets entirely

**Comment by Gillibald** [comment]
To make it clear we don't need CharacterBufferReference to be able to port something from WPF. 

Hit testing code still works with ReadOnlySpan. So the rework was still needed but API parity was wrong.

**Comment by MrJul** [comment]
Thank you all for your answers and clarifications, much appreciated!

I'm glad performance is prioritized over compatibility, and that old WPF types aren't mandatory.
I'll make a PR for proper review.
Edit: opened PR #9982
